### ISSUE TITLE
Closing a pop up window causing geckodriver to throw WebDriverException

### ISSUE THREAD

**Issue by JoshuaBurns** [mozilla/geckodriver#756]
In order to help us efficiently investigate your issue, please provide the following information:

## Firefox Version
Firefox Nightly 55.0a1

## Platform
Windows 7 professional
## Steps to reproduce -
The issue happens when I close a pop up window the test fails with a `The HTTP request to the remote WebDriver server for URL http://localhost:4444/session/94fc3bec-4895-484e-89b8-8c3a30bf3c24/element/7e5ff9c6-a267-4c4d-b511-86f9e59f3a51/click timed out after 30 seconds.` which as I understand it means webdriver sent out a request and got nothing back and thus didnt know how to proceed. 

The test I provided is using 

```
  <package id="log4net" version="2.0.8" targetFramework="net45" />
  <package id="NUnit" version="3.6.1" targetFramework="net45" />
  <package id="NUnit.ConsoleRunner" version="3.6.1" targetFramework="net45" />
  <package id="NUnit.Extension.NUnitProjectLoader" version="3.5.0" targetFramework="net45" />
  <package id="NUnit.Extension.NUnitV2Driver" version="3.6.0" targetFramework="net45" />
  <package id="NUnit.Extension.NUnitV2ResultWriter" version="3.5.0" targetFramework="net45" />
  <package id="NUnit.Extension.TeamCityEventListener" version="1.0.2" targetFramework="net45" />
  <package id="NUnit.Extension.VSProjectLoader" version="3.5.0" targetFramework="net45" />
  <package id="NUnit.Runners" version="3.6.1" targetFramework="net45" />
  <package id="NUnit3TestAdapter" version="3.7.0" targetFramework="net45" />
  <package id="Selenium.Support" version="3.4.0" targetFramework="net45" />
  <package id="Selenium.WebDriver" version="3.4.0" targetFramework="net45" />
  <package id="Selenium.WebDriver.ChromeDriver" version="2.29.0" targetFramework="net45" />
  <package id="Selenium.WebDriver.GeckoDriver" version="0.16.1" targetFramework="net45" />
  <package id="Selenium.WebDriver.GeckoDriver.Win32" version="0.16.1" targetFramework="net45" />
  <package id="Selenium.WebDriver.IEDriver" version="3.4.0" targetFramework="net45" />
```

```
using NUnit.Framework;
using OpenQA.Selenium;
using OpenQA.Selenium.Firefox;
using System;
using OpenQA.Selenium.Support.UI;

namespace POMAuctivaTest.TestSuite
{
    [TestFixture]
    [Category("test")]
    public class Test
    {
        public IWebDriver driver { get; set; }
        [OneTimeSetUp]
        public void setUp()
        {
            FirefoxDriverService service = FirefoxDriverService.CreateDefaultService();
            service.Port = 4444;
            FirefoxOptions ffoptions = new FirefoxOptions();
            ffoptions.BrowserExecutableLocation = @"C:\Program Files\Nightly\firefox.exe";
            ffoptions.SetLoggingPreference(LogType.Browser.ToString(), LogLevel.All);
            ffoptions.LogLevel = FirefoxDriverLogLevel.Trace;
            ICapabilities caps = ffoptions.ToCapabilities();
            driver = new FirefoxDriver(service, ffoptions, TimeSpan.FromSeconds(30));
        }

        [Test]
        public void test()
        {
            driver.Manage().Window.Maximize();
            driver.Navigate().GoToUrl("https://www.auctiva.com/default.aspx");
            WaitForElementVisible(driver, By.CssSelector("#ftctrl_header_ftctrl_SignIn_btn_SignIn"));
            driver.FindElement(By.CssSelector("#ftctrl_header_ftctrl_SignIn_txt_Username")).SendKeys("GeckoDriverTest");
            driver.FindElement(By.CssSelector("#ftctrl_header_ftctrl_SignIn_txt_Password")).SendKeys("marionette`");
            driver.FindElement(By.CssSelector("#ftctrl_header_ftctrl_SignIn_btn_SignIn")).Click();
            WaitForElementVisible(driver, By.CssSelector("#btctrl_Header_lnk_signout"));
            driver.Navigate().GoToUrl("https://www.auctiva.com/listings/createlisting.aspx");
            WaitForElementVisible(driver, By.CssSelector("#ftctrl_DetailSection_dl_Images__ctl0_img_Image"));
            string currentHandle = driver.CurrentWindowHandle;
            PopupWindowFinder finder = new PopupWindowFinder(driver, TimeSpan.FromSeconds(80));
            var popupWindowHandle = finder.Click(driver.FindElement(By.CssSelector("#ftctrl_DetailSection_dl_Images__ctl0_img_Image")));
            driver.SwitchTo().Window(popupWindowHandle);
            WaitForElementVisible(driver, By.CssSelector("#imageSelectorThumbs_rpt_ImageList__ctl1_img_img"));
            driver.FindElement(By.CssSelector("#imageSelectorThumbs_rpt_ImageList__ctl1_img_img")).Click();
            driver.FindElement(By.CssSelector("#btn_done2")).Click(); //  This is where the issue occurs
            driver.SwitchTo().Window(currentHandle);  // Not executed
            WaitForElementVisible(driver, By.CssSelector("#Header_lnk_signout")); // Not executed
        }

        [OneTimeTearDown]
        public void tearDown()
        {
            driver.Quit();
        }

        public void WaitForElementVisible(IWebDriver driver, By element)
        {
            var attempts = 0;
            while (attempts < 2)
            {
                try
                {
                    WebDriverWait wait = new WebDriverWait(driver, TimeSpan.FromSeconds(15));
                    wait.PollingInterval = TimeSpan.FromMilliseconds(300);
                    wait.Until(ExpectedConditions.ElementIsVisible(element));
                    break;
                }
                catch (Exception)
                {
                    attempts++;
                }
            }
            if (attempts >= 2)
            {
                throw new NoSuchElementException("Cannot locate " + element.ToString());
            }
        }
    }
}

```

Attaching two trace logs `test.txt` I gathered using `Out-file C:\test.txt` but this seemed incomplete as I didnt notice the same ending log entrees that I saw in the cmd window so `test1.txt` is from a copy and past from cmd which is why it is missing the start of the test. 
[test.txt](https://github.com/mozilla/geckodriver/files/1029990/test.txt)
[test1.txt](https://github.com/mozilla/geckodriver/files/1029991/test1.txt)

** If any of the above are missing we will have to unfortunately close your issue. We will gladly reopen the issue once all the information requested has been added **

**Comment by andreastt** [comment]
Are you able to come up with a more minimised test case? I don’t have the time right now to dig through log files with thousands of lines.

**Comment by JoshuaBurns** [comment]
I am trying to find a more minimised test case, currently unable have not found one. I will keep looking but verified that the same issue is consistently happening on the long test case I originally posted.

**Comment by whimboo** [comment]
I'm sure this relies on issue #611, which will be fixed soon in Firefox Nightly.

**Comment by JoshuaBurns** [comment]
Here is some additional info I have noticed in regards to this issue. 
Tested with the above test case and the following configurations.

This IS NOT an issue with Firefox 53.0.3 and GeckDriver 0.16.1

This IS an issue with Nightly 55.0a1 and GeckoDriver 0.16.1
This IS an issue with Firefox 53.0.3 and GeckoDriver 0.17.0 
This IS an issue with Firefox 54.0 and GeckoDriver 0.17.0

**Comment by whimboo** [comment]
I just read the issue carefully again and noticed that in the original comment a hang was reported when issuing a click which closes a popup. So this is what has been fixed lately and also got reported as issue #611. Closing this one as dupe.
